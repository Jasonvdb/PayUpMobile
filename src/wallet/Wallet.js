import {
	formattedTransactionsFromAddresses,
	generateMnemonic,
	getAddressFromXpub,
	getXpubFromMnemonic
} from "./wallet-functions";
import { getAddressBalance, getAddressTransactions } from "./blockexplorer";

const bitcoin = require("bitcoinjs-lib");

const networks = {
	mainnet: {
		network: bitcoin.networks.bitcoin,
		apiBaseUrl: "https://blockstream.info/api/"
	},
	testnet: {
		network: bitcoin.networks.testnet,
		apiBaseUrl: "https://blockstream.info/testnet/api/"
	}
};

/**
 * HD Wallet (BIP39).
 * BIP49 (P2SH Segwit)
 * @see https://github.com/bitcoin/bips/blob/master/bip-0049.mediawiki
 */
export default class Wallet {
  network = null;

  mnemonic = null;

  xpub = null;

  receiveAddresses = [];

  changeAddresses = [];

  apiBaseUrl = null;

  addressUpdateQueue = [];

  rawTransactions = {};

  neatTransactionHistory = [];

  addressBalances = {};

  constructor(networkName) {
  	if (!networkName || !networks[networkName]) {
  		throw new Error(
  			`Please specify a valid network (${Object.keys(networks).join(", ")})`
  		);
  	}
  	const { network, apiBaseUrl } = networks[networkName];
  	this.network = network;
  	this.apiBaseUrl = apiBaseUrl;
  }

  async createNewWallet() {
  	const mnemonic = await generateMnemonic();
  	this._init(mnemonic);
  }

  importExistingWallet(mnemonic) {
  	this._init(mnemonic);
  }

  _init(mnemonic) {
  	if (this.mnemonic) {
  		throw new Error("Existing wallet already loaded.");
  	}

  	this.mnemonic = mnemonic.trim();
  	this.xpub = getXpubFromMnemonic(this.mnemonic, this.network);
  	this.appendDerivedAddresses(20);
  }

  appendDerivedAddresses(numberOfAddresses) {
  	if (!this.xpub) {
  		throw new Error("No wallet loaded.");
  	}

  	const startIndex = this.receiveAddresses.length;
  	const endIndex = startIndex + numberOfAddresses;

  	for (let index = startIndex; index < endIndex; index++) {
  		this.receiveAddresses.push(
  			getAddressFromXpub(this.xpub, index, this.network, "receive")
  		);

  		this.changeAddresses.push(
  			getAddressFromXpub(this.xpub, index, this.network, "change")
  		);
  	}
  }

  queueAddressForUpdate(address) {
  	this.addressUpdateQueue.push(address);
  	this.processAddressTransactionsFromQueue();
  }

  processAddressTransactionsFromQueue() {
  	//Busy processing already OR nothing to process
  	if (this.busyUpdatingAddressTxData) {
  		return;
  	}

  	if (this.addressUpdateQueue.length === 0) {
  		this.busyUpdatingAddressTxData = false;
  	}

  	const address = this.addressUpdateQueue[0];

  	this.busyUpdatingAddressTxData = true;
  	this.updateAddressTransactions(address)
  		.then(() => {
  			this.addressUpdateQueue.shift();
  		})
  		.catch(e => {
  			throw e;
  		})
  		.then(() => {
  			this.busyUpdatingAddressTxData = false;
  			this.processAddressTransactionsFromQueue();
  		});
  }

  async updateAddressTransactions(address) {
  	const rawTransactions = await getAddressTransactions(
  		this.apiBaseUrl,
  		address
  	);

  	rawTransactions.forEach(tx => {
  		const { txid } = tx;
  		this.rawTransactions[txid] = { ...tx, refreshedAt: new Date() };
  	});
  }

  //Used to poll when the user is waiting for funds
  async updateAddressBalance(address) {
  	const {
  		confirmedValueInSats,
  		unconfirmedValueInSats
  	} = await getAddressBalance(this.apiBaseUrl, address);

  	//If there's a confirmed balance, update the rest of the wallet
  	if (confirmedValueInSats > 0) {
  		await this.updateAddressTransactions(address);
  		this.updateTransactionHistory();
  	}

  	//TODO call updateAddressTransactions and then updateTransactionHistory

  	this.addressBalances[address] = {
  		confirmedValueInSats,
  		unconfirmedValueInSats,
  		updatedAt: new Date()
  	};
  }

  updateTransactionHistory() {
  	this.neatTransactionHistory = formattedTransactionsFromAddresses(
  		this.rawTransactions,
  		this.receiveAddresses,
  		this.changeAddresses
  	);
  }
}
