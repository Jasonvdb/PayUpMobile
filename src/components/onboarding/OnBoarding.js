import React, { Component, Fragment } from "react";
import { StyleSheet, View, StatusBar, Dimensions } from "react-native";
import SnapCarousel from "react-native-snap-carousel";
import { inject, observer } from "mobx-react";
import PropTypes from "prop-types";

import theme from "../../config/theme";
import Slide from "./Slide";
import onError from "../../helpers/onError";
import Wallet from "../../wallet/Wallet";
const windowWidth = Dimensions.get("window").width;

const sliderWidth = windowWidth;

class OnBoarding extends Component {
  constructor(props) {
    super(props);

    this.state = {
      isCreating: false,
      isImporting: false
    };
  }

  componentDidMount(): void {}

  snapToNext() {
    setTimeout(() => this._carousel.snapToNext(), 50);
  }

  snapToPrevious() {
    setTimeout(() => this._carousel.snapToPrev(), 50);
  }

  onCreate() {
    //TODO ask for testnet or mainnet
    this.setState({ isCreating: true });
    this.props.wallet
      .createNewWallet()
      .then(() => {})
      .catch(e => {
        onError(e, "Failed to create new wallet.");

        this.setState({ isCreating: false });
      });
  }

  onImport() {
    this.setState({ isImporting: true });
    alert("Import");
  }

  render() {
    const { isCreating, isImporting } = this.state;

    const items = [
      {
        imageSource: require("../../../images/onboarding_2.png"),
        title: "A Bitcoin wallet betting app",
        content:
          "A bitcoin wallet with the added feature of allowing you to make secure bets with friends."
      },
      {
        imageSource: require("../../../images/onboarding_5.png"),
        title: "Make bets with your friends",
        content:
          "Add funds to the wallet, invite some friends and start betting. "
      },
      {
        imageSource: require("../../../images/onboarding_3.png"),
        title: "Your keys, your money",
        content:
          "Control of your funds can only be done on the app, and funds pending a bet outcome are controlled jointly by you and your friend."
      },
      {
        imageSource: require("../../../images/onboarding_6.png"),
        title: "Get started",
        content:
          "Either setup your wallet for the first time or import one from a backup."
      }
    ];

    return (
      <Fragment>
        <StatusBar barStyle="light-content"/>
        <View style={styles.root}>
          <SnapCarousel
            layoutCardOffset={30}
            inactiveSlideOpacity={1}
            inactiveSlideScale={1}
            ref={c => {
              this._carousel = c;
            }}
            data={items}
            renderItem={props => {
              const { index } = props;

              //Every slide except first needs a back button
              if (index !== 0) {
                props.onBackPress = this.snapToPrevious.bind(this);
              }

              //Every slide except last needs continue button
              if (index !== 3) {
                props.onContinuePress = this.snapToNext.bind(this);
              }

              //Last slide
              if (index === 3) {
                props.onCreate = this.onCreate.bind(this);
                props.onImport = this.onImport.bind(this);
              }

              return (
                <Slide
                  {...props}
                  isCreating={isCreating}
                  isImporting={isImporting}
                />
              );
            }}
            sliderWidth={sliderWidth}
            itemWidth={sliderWidth}
          />
        </View>
      </Fragment>
    );
  }
}

OnBoarding.propTypes = {
  wallet: PropTypes.instanceOf(Wallet)
};

const styles = StyleSheet.create({
  root: {
    height: "100%",
    backgroundColor: theme.background
  }
});

export default inject("wallet")(observer(OnBoarding));
