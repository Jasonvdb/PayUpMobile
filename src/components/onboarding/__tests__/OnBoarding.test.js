import React from "react";
import { shallow, render } from "enzyme";
import OnBoarding from "../OnBoarding";
import Wallet from "../../../wallet/Wallet";

describe("on boarding", () => {
  const origErrorConsole = console.error;
  const origWarnConsole = console.warn;
  beforeEach(() => {
    console.error = message => {
      if (message.indexOf("Use PascalCase" > -1)) {
        return;
      }

      origErrorConsole(message);
    };

    console.warn = message => {
      if (message.indexOf("Animated:" > -1)) {
        return;
      }

      origWarnConsole(message);
    };
  });

  afterEach(() => {
    console.error = origErrorConsole;
    console.warn = origWarnConsole;
  });

  describe("Rendering", () => {
    let wallet;
    beforeAll(() => {
      wallet = new Wallet();
    });

    it("should match to snapshot - on boarding container", () => {
      const component = render(<OnBoarding wallet={wallet}/>);
      expect(component).toMatchSnapshot("On boarding container snapshot");
    });
  });

  // describe("interaction", () => {
  //
  // });
});
