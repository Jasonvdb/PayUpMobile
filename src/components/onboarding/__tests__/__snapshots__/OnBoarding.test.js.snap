// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`on boarding Rendering should match to snapshot - on boarding container: On boarding container snapshot 1`] = `
initialize {
  "0": Object {
    "attribs": Object {
      "style": "height:100%;background-color:#f8f8f8",
    },
    "children": Array [
      Object {
        "attribs": Object {
          "activeanimationtype": "timing",
          "activeslidealignment": "center",
          "activeslideoffset": "20",
          "apparitiondelay": "0",
          "autoplaydelay": "1000",
          "autoplayinterval": "3000",
          "callbackoffsetmargin": "5",
          "containercustomstyle": "[object Object]",
          "contentcontainercustomstyle": "[object Object]",
          "contentcontainerstyle": "[object Object],[object Object]",
          "data": "[object Object],[object Object],[object Object],[object Object]",
          "decelerationrate": "fast",
          "firstitem": "0",
          "inactiveslideopacity": "1",
          "inactiveslidescale": "1",
          "inactiveslideshift": "0",
          "initialnumtorender": "6",
          "itemwidth": "750",
          "layout": "default",
          "layoutcardoffset": "30",
          "lockscrolltimeoutduration": "1000",
          "loopclonesperside": "3",
          "maxtorenderperbatch": "13",
          "numcolumns": "1",
          "overscrollmode": "never",
          "scrolleventthrottle": "1",
          "sliderwidth": "750",
          "slidestyle": "[object Object]",
          "stickyheaderindices": "",
          "style": "opacity:0;width:750px;flex-direction:row",
          "swipethreshold": "20",
          "updatecellsbatchingperiod": "50",
          "viewabilityconfigcallbackpairs": "",
          "windowsize": "13",
        },
        "children": Array [
          Object {
            "attribs": Object {},
            "children": Array [
              Object {
                "attribs": Object {
                  "style": "0:[object Object]",
                },
                "children": Array [],
                "name": "view",
                "namespace": "http://www.w3.org/1999/xhtml",
                "next": Object {
                  "attribs": Object {
                    "style": "0:[object Object]",
                  },
                  "children": Array [],
                  "name": "view",
                  "namespace": "http://www.w3.org/1999/xhtml",
                  "next": Object {
                    "attribs": Object {
                      "style": "0:[object Object]",
                    },
                    "children": Array [],
                    "name": "view",
                    "namespace": "http://www.w3.org/1999/xhtml",
                    "next": Object {
                      "attribs": Object {
                        "style": "0:[object Object]",
                      },
                      "children": Array [],
                      "name": "view",
                      "namespace": "http://www.w3.org/1999/xhtml",
                      "next": null,
                      "parent": [Circular],
                      "prev": [Circular],
                      "type": "tag",
                      "x-attribsNamespace": Object {
                        "style": undefined,
                      },
                      "x-attribsPrefix": Object {
                        "style": undefined,
                      },
                    },
                    "parent": [Circular],
                    "prev": [Circular],
                    "type": "tag",
                    "x-attribsNamespace": Object {
                      "style": undefined,
                    },
                    "x-attribsPrefix": Object {
                      "style": undefined,
                    },
                  },
                  "parent": [Circular],
                  "prev": [Circular],
                  "type": "tag",
                  "x-attribsNamespace": Object {
                    "style": undefined,
                  },
                  "x-attribsPrefix": Object {
                    "style": undefined,
                  },
                },
                "parent": [Circular],
                "prev": null,
                "type": "tag",
                "x-attribsNamespace": Object {
                  "style": undefined,
                },
                "x-attribsPrefix": Object {
                  "style": undefined,
                },
              },
              Object {
                "attribs": Object {
                  "style": "0:[object Object]",
                },
                "children": Array [],
                "name": "view",
                "namespace": "http://www.w3.org/1999/xhtml",
                "next": Object {
                  "attribs": Object {
                    "style": "0:[object Object]",
                  },
                  "children": Array [],
                  "name": "view",
                  "namespace": "http://www.w3.org/1999/xhtml",
                  "next": Object {
                    "attribs": Object {
                      "style": "0:[object Object]",
                    },
                    "children": Array [],
                    "name": "view",
                    "namespace": "http://www.w3.org/1999/xhtml",
                    "next": null,
                    "parent": [Circular],
                    "prev": [Circular],
                    "type": "tag",
                    "x-attribsNamespace": Object {
                      "style": undefined,
                    },
                    "x-attribsPrefix": Object {
                      "style": undefined,
                    },
                  },
                  "parent": [Circular],
                  "prev": [Circular],
                  "type": "tag",
                  "x-attribsNamespace": Object {
                    "style": undefined,
                  },
                  "x-attribsPrefix": Object {
                    "style": undefined,
                  },
                },
                "parent": [Circular],
                "prev": Object {
                  "attribs": Object {
                    "style": "0:[object Object]",
                  },
                  "children": Array [],
                  "name": "view",
                  "namespace": "http://www.w3.org/1999/xhtml",
                  "next": [Circular],
                  "parent": [Circular],
                  "prev": null,
                  "type": "tag",
                  "x-attribsNamespace": Object {
                    "style": undefined,
                  },
                  "x-attribsPrefix": Object {
                    "style": undefined,
                  },
                },
                "type": "tag",
                "x-attribsNamespace": Object {
                  "style": undefined,
                },
                "x-attribsPrefix": Object {
                  "style": undefined,
                },
              },
              Object {
                "attribs": Object {
                  "style": "0:[object Object]",
                },
                "children": Array [],
                "name": "view",
                "namespace": "http://www.w3.org/1999/xhtml",
                "next": Object {
                  "attribs": Object {
                    "style": "0:[object Object]",
                  },
                  "children": Array [],
                  "name": "view",
                  "namespace": "http://www.w3.org/1999/xhtml",
                  "next": null,
                  "parent": [Circular],
                  "prev": [Circular],
                  "type": "tag",
                  "x-attribsNamespace": Object {
                    "style": undefined,
                  },
                  "x-attribsPrefix": Object {
                    "style": undefined,
                  },
                },
                "parent": [Circular],
                "prev": Object {
                  "attribs": Object {
                    "style": "0:[object Object]",
                  },
                  "children": Array [],
                  "name": "view",
                  "namespace": "http://www.w3.org/1999/xhtml",
                  "next": [Circular],
                  "parent": [Circular],
                  "prev": Object {
                    "attribs": Object {
                      "style": "0:[object Object]",
                    },
                    "children": Array [],
                    "name": "view",
                    "namespace": "http://www.w3.org/1999/xhtml",
                    "next": [Circular],
                    "parent": [Circular],
                    "prev": null,
                    "type": "tag",
                    "x-attribsNamespace": Object {
                      "style": undefined,
                    },
                    "x-attribsPrefix": Object {
                      "style": undefined,
                    },
                  },
                  "type": "tag",
                  "x-attribsNamespace": Object {
                    "style": undefined,
                  },
                  "x-attribsPrefix": Object {
                    "style": undefined,
                  },
                },
                "type": "tag",
                "x-attribsNamespace": Object {
                  "style": undefined,
                },
                "x-attribsPrefix": Object {
                  "style": undefined,
                },
              },
              Object {
                "attribs": Object {
                  "style": "0:[object Object]",
                },
                "children": Array [],
                "name": "view",
                "namespace": "http://www.w3.org/1999/xhtml",
                "next": null,
                "parent": [Circular],
                "prev": Object {
                  "attribs": Object {
                    "style": "0:[object Object]",
                  },
                  "children": Array [],
                  "name": "view",
                  "namespace": "http://www.w3.org/1999/xhtml",
                  "next": [Circular],
                  "parent": [Circular],
                  "prev": Object {
                    "attribs": Object {
                      "style": "0:[object Object]",
                    },
                    "children": Array [],
                    "name": "view",
                    "namespace": "http://www.w3.org/1999/xhtml",
                    "next": [Circular],
                    "parent": [Circular],
                    "prev": Object {
                      "attribs": Object {
                        "style": "0:[object Object]",
                      },
                      "children": Array [],
                      "name": "view",
                      "namespace": "http://www.w3.org/1999/xhtml",
                      "next": [Circular],
                      "parent": [Circular],
                      "prev": null,
                      "type": "tag",
                      "x-attribsNamespace": Object {
                        "style": undefined,
                      },
                      "x-attribsPrefix": Object {
                        "style": undefined,
                      },
                    },
                    "type": "tag",
                    "x-attribsNamespace": Object {
                      "style": undefined,
                    },
                    "x-attribsPrefix": Object {
                      "style": undefined,
                    },
                  },
                  "type": "tag",
                  "x-attribsNamespace": Object {
                    "style": undefined,
                  },
                  "x-attribsPrefix": Object {
                    "style": undefined,
                  },
                },
                "type": "tag",
                "x-attribsNamespace": Object {
                  "style": undefined,
                },
                "x-attribsPrefix": Object {
                  "style": undefined,
                },
              },
            ],
            "name": "view",
            "namespace": "http://www.w3.org/1999/xhtml",
            "next": null,
            "parent": [Circular],
            "prev": null,
            "type": "tag",
            "x-attribsNamespace": Object {},
            "x-attribsPrefix": Object {},
          },
        ],
        "name": "rctscrollview",
        "namespace": "http://www.w3.org/1999/xhtml",
        "next": null,
        "parent": [Circular],
        "prev": null,
        "type": "tag",
        "x-attribsNamespace": Object {
          "activeanimationtype": undefined,
          "activeslidealignment": undefined,
          "activeslideoffset": undefined,
          "apparitiondelay": undefined,
          "autoplaydelay": undefined,
          "autoplayinterval": undefined,
          "callbackoffsetmargin": undefined,
          "containercustomstyle": undefined,
          "contentcontainercustomstyle": undefined,
          "contentcontainerstyle": undefined,
          "data": undefined,
          "decelerationrate": undefined,
          "firstitem": undefined,
          "inactiveslideopacity": undefined,
          "inactiveslidescale": undefined,
          "inactiveslideshift": undefined,
          "initialnumtorender": undefined,
          "itemwidth": undefined,
          "layout": undefined,
          "layoutcardoffset": undefined,
          "lockscrolltimeoutduration": undefined,
          "loopclonesperside": undefined,
          "maxtorenderperbatch": undefined,
          "numcolumns": undefined,
          "overscrollmode": undefined,
          "scrolleventthrottle": undefined,
          "sliderwidth": undefined,
          "slidestyle": undefined,
          "stickyheaderindices": undefined,
          "style": undefined,
          "swipethreshold": undefined,
          "updatecellsbatchingperiod": undefined,
          "viewabilityconfigcallbackpairs": undefined,
          "windowsize": undefined,
        },
        "x-attribsPrefix": Object {
          "activeanimationtype": undefined,
          "activeslidealignment": undefined,
          "activeslideoffset": undefined,
          "apparitiondelay": undefined,
          "autoplaydelay": undefined,
          "autoplayinterval": undefined,
          "callbackoffsetmargin": undefined,
          "containercustomstyle": undefined,
          "contentcontainercustomstyle": undefined,
          "contentcontainerstyle": undefined,
          "data": undefined,
          "decelerationrate": undefined,
          "firstitem": undefined,
          "inactiveslideopacity": undefined,
          "inactiveslidescale": undefined,
          "inactiveslideshift": undefined,
          "initialnumtorender": undefined,
          "itemwidth": undefined,
          "layout": undefined,
          "layoutcardoffset": undefined,
          "lockscrolltimeoutduration": undefined,
          "loopclonesperside": undefined,
          "maxtorenderperbatch": undefined,
          "numcolumns": undefined,
          "overscrollmode": undefined,
          "scrolleventthrottle": undefined,
          "sliderwidth": undefined,
          "slidestyle": undefined,
          "stickyheaderindices": undefined,
          "style": undefined,
          "swipethreshold": undefined,
          "updatecellsbatchingperiod": undefined,
          "viewabilityconfigcallbackpairs": undefined,
          "windowsize": undefined,
        },
      },
    ],
    "name": "view",
    "namespace": "http://www.w3.org/1999/xhtml",
    "next": null,
    "parent": null,
    "prev": null,
    "root": Object {
      "attribs": Object {},
      "children": Array [
        [Circular],
      ],
      "name": "root",
      "namespace": "http://www.w3.org/1999/xhtml",
      "next": null,
      "parent": null,
      "prev": null,
      "type": "root",
      "x-attribsNamespace": Object {},
      "x-attribsPrefix": Object {},
    },
    "type": "tag",
    "x-attribsNamespace": Object {
      "style": undefined,
    },
    "x-attribsPrefix": Object {
      "style": undefined,
    },
  },
  "_root": [Circular],
  "length": 1,
  "options": Object {
    "decodeEntities": true,
    "normalizeWhitespace": false,
    "withDomLvl1": true,
    "xml": false,
  },
}
`;
